BUG ONNXImporter::handleNode MatMul

### System Information

OpenCV version: current Master branch 
Operating System / Platform: Windows 11
Compiler & compiler version: MSVC 16


### Detailed description

When trying to parse the onnx file (exported from https://github.com/baudm/parseq and simplified with https://github.com/daquexian/onnx-simplifier as recommended in the opencv doku) i get following error:

> ERROR:0@0.145] global X:\ThirdPartySource\opencv_cuda\opencv\modules\dnn\src\onnx\onnx_importer.cpp (1050) cv::dnn::dnn4_v20220524::ONNXImporter::handleNode DNN/ONNX: ERROR during processing node with 2 inputs and 1 outputs: [MatMul]:(onnx_node!MatMul_859) from domain='ai.onnx'
OpenCV: terminate handler is called! The last OpenCV error is:
OpenCV(4.6.0-dev) Error: Unspecified error (> Node [MatMul@ai.onnx]:(onnx_node!MatMul_859) parse error: OpenCV(4.6.0-dev) X:\ThirdPartySource\opencv_cuda\opencv\modules\core\src\matrix_transform.cpp:245: error: (-215:Assertion failed) _src.dims() <= 2 && esz <= 32 in function 'cv::transpose'
> ) in cv::dnn::dnn4_v20220524::ONNXImporter::handleNode, file X:\ThirdPartySource\opencv_cuda\opencv\modules\dnn\src\onnx\onnx_importer.cpp, line 1069

The onnx it self is valid and can be used with the onnx runtime itself.
i also checked it via https://netron.app/

there seems to be 2 issues in MatMul_859 first i think opencv expect another order of the inputs.
But even if i change them i get another error here i am :)



### Steps to reproduce

auto detector = cv::dnn::readNetFromONNX("ocr-sim.onnx");


weetransfer for the onnx file:

https://wetransfer.com/downloads/46d0dc6287c0d1ed000924a036400bbe20221028095359/f7770d3acc028f8006a96bced82790c920221028095418/e4e0a6



### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

--- a/modules/dnn/src/onnx/onnx_importer.cpp
+++ b/modules/dnn/src/onnx/onnx_importer.cpp
@@ -2031,8 +2031,9 @@ void ONNXImporter::parseGemm(LayerParams& layerParams, const opencv_onnx::NodePr
     addLayer(layerParams, node_proto);
 }
 
-void ONNXImporter::parseMatMul(LayerParams& layerParams, const opencv_onnx::NodeProto& node_proto)
+void ONNXImporter::parseMatMul(LayerParams& layerParams, const opencv_onnx::NodeProto& node_proto_)
 {
+    opencv_onnx::NodeProto node_proto = node_proto_;
     CV_Assert(node_proto.input_size() == 2);
     layerParams.type = "InnerProduct";
     layerParams.set("bias_term", false);
@@ -2044,8 +2045,24 @@ void ONNXImporter::parseMatMul(LayerParams& layerParams, const opencv_onnx::Node
     {
         Mat blob = getBlob(node_proto, 1);
         secondInpDims = blob.dims;
-        layerParams.blobs.push_back(blob.t());
-        layerParams.set("num_output", layerParams.blobs[0].size[0]);
+        if (secondInpDims == 2)
+        {
+            layerParams.blobs.push_back(blob.t());
+            layerParams.set("num_output", layerParams.blobs[0].size[0]);
+        }
+        else
+        {
+            LayerParams constParams;
+            constParams.name = layerParams.name + "/const";
+            constParams.type = "Const";
+            constParams.blobs.push_back(blob);
+
+            opencv_onnx::NodeProto tmpProto;
+            tmpProto.add_output(constParams.name);
+            addLayer(constParams, tmpProto);
+
+            node_proto.set_input(1, constParams.name);
+        }
     } else {
         secondInpDims = outShapes[node_proto.input(1)].size();
     }

--- a/modules/dnn/test/test_onnx_importer.cpp
+++ b/modules/dnn/test/test_onnx_importer.cpp
@@ -956,6 +956,13 @@ TEST_P(Test_ONNX_layers, MatMul)
     testONNXModels("matmul_4d");
 }
 
+TEST_P(Test_ONNX_layers, MatMul_init)
+{
+    testONNXModels("matmul_2d_init");
+    testONNXModels("matmul_3d_init");
+    testONNXModels("matmul_4d_init");
+}
+
 TEST_P(Test_ONNX_layers, MatMulAdd)
 {
 #if defined(INF_ENGINE_RELEASE) && INF_ENGINE_VER_MAJOR_EQ(2022010000)

